tensorboard/lazy.py

```diff
@@ -19,36 +19,63 @@
 from __future__ import division
 from __future__ import print_function
 
-import importlib
+import functools
+import threading
 import types
 
 
-class LazyLoader(types.ModuleType):
-  """Lazily import a module, mainly to avoid pulling in large dependencies."""
+def lazy_load(name):
+  """Decorator to define a function that lazily loads the module 'name'.
 
-  # The lint error here is incorrect.
-  def __init__(self, local_name, parent_module_globals, name):  # pylint: disable=super-on-old-class
-    self._local_name = local_name
-    self._parent_module_globals = parent_module_globals
+  This can be used to defer importing troublesome dependencies - e.g. ones that
+  are large and infrequently used, or that cause a dependency cycle -
+  until they are actually used.
 
-    super(LazyLoader, self).__init__(name)
+  Args:
+    name: the fully-qualified name of the module; typically the last segment
+      of 'name' matches the name of the decorated function
 
-  def _load(self):
-    # Import the target module and insert it into the parent's namespace
-    module = importlib.import_module(self.__name__)
-    self._parent_module_globals[self._local_name] = module
+  Returns:
+    Decorator function that produces a lazy-loading module 'name' backed by the
+    underlying decorated function.
+  """
+  def wrapper(load_fn):
+    # Wrap load_fn to call it exactly once and update __dict__ afterwards to
+    # make future lookups efficient (only failed lookups call __getattr__).
+    @_return_once
+    def load_once(self):
+      module = load_fn()
+      self.__dict__.update(module.__dict__)
+      return module
 
-    # Update this object's dict so that if someone keeps a reference to the
-    #   LazyLoader, lookups are efficient (__getattr__ is only called on lookups
-    #   that fail).
-    self.__dict__.update(module.__dict__)
+    # Define a module that proxies getattr() and dir() to the result of calling
+    # load_once() the first time it's needed. The class is nested so we can close
+    # over load_once() and avoid polluting the module's attrs with our own state.
+    class LazyModule(types.ModuleType):
+      def __getattr__(self, attr_name):
+        return getattr(load_once(self), attr_name)
 
-    return module
+      def __dir__(self):
+        return dir(load_once(self))
 
-  def __getattr__(self, item):
-    module = self._load()
-    return getattr(module, item)
+      def __repr__(self):
+        return '<module \'%s\' (LazyModule)>' % self.__name__
 
-  def __dir__(self):
-    module = self._load()
-    return dir(module)
+    return LazyModule(name)
+  return wrapper
+
+
+def _return_once(f):
+  """Decorator that calls f() once, then returns that value repeatedly."""
```

Comment:
Though used safely\* within this module, this decorator seems dangerous in
general; it passes through arguments to `f` the first time, then
completely ignores them at all subsequent invocations because the cached
result already exists.

Given that the local uses are safe, I’d be fine with either of:

  - documenting this with a warning, and changing the headline of the
    docstring from `calls f()` to `` calls `f` ``; or

  - refactoring so that the decorator actually only calls `f()` (no
    args), and the `self` argument to `load_once` is passed in
    separately:

    ```python
    lazy_module_box = []

    def load_once():
      module = load_fn()
      lazy_module_box[0].__dict__.update(module.__dict__)
      return module

    class LazyModule(types.ModuleType):
      # …
      pass

    lazy_module = LazyModule(name)
    lazy_module_box.append(lazy_module)
    return lazy_module
    ```

(You could also remove the decorator altogether, but I agree that it’s
nice to have the locking logic encapsulated.)

\* Barring clients who do things like `type(mod).__dir__(not_mod)`,
which I think is fine to ignore.

Reply:
It was meant to just be an implementation detail so I wasn't worrying about this, but fair enough.  I rewrote the decorator to be a standard `memoize` that only supports a single hashable argument since that's all that we need.